ooni/nettests/blocking/web_connectivity.py

```diff
@@ -0,0 +1,512 @@
+            return True
+
+        control_ips = set(self.control['dns']['ips'])
+        experiment_ips = set(experiment_dns_answers)
+
+        if control_ips == experiment_ips:
+            return True
+
+        for experiment_ip in experiment_ips:
+            if is_public_ipv4_address(experiment_ip) is False:
+                return False
+
+        if len(control_ips.intersection(experiment_ips)) > 0:
+            return True
+
+        experiment_asns = set(map(lambda x: geoip.IPToLocation(x)['asn'],
+                              experiment_ips))
+        control_asns = set(map(lambda x: geoip.IPToLocation(x)['asn'],
+                           control_ips))
+
+        # Remove the instance of AS0 when we fail to find the ASN
+        control_asns.discard('AS0')
+        experiment_asns.discard('AS0')
+
+        if len(control_asns.intersection(experiment_asns)) > 0:
+            return True
+
+        return False
+
+    def compare_tcp_experiments(self):
+        success = True
+        for idx, result in enumerate(self.report['tcp_connect']):
+            socket = "%s:%s" % (result['ip'], result['port'])
+            control_status = self.control['tcp_connect'][socket]
+            if result['status']['success'] == False and \
+                    control_status['status'] == True:
+                self.report['tcp_connect'][idx]['status']['blocked'] = True
+                success = False
+            else:
+                self.report['tcp_connect'][idx]['status']['blocked'] = False
+        return success
+
+    def determine_blocking(self, experiment_http_response, experiment_dns_answers):
+        blocking = False
+
+        control_http_failure = self.control['http_request']['failure']
+        if control_http_failure is not None:
+            control_http_failure = control_http_failure.split(" ")[0]
+
+        experiment_http_failure = self.report['http_experiment_failure']
+        if experiment_http_failure is not None:
+            experiment_http_failure = experiment_http_failure.split(" ")[0]
+
+        if (experiment_http_failure is None and control_http_failure is None):
+            self.compare_http_experiments(experiment_http_response)
+
+        dns_consistent = self.compare_dns_experiments(experiment_dns_answers)
+        if dns_consistent is True:
+            self.report['dns_consistency'] = 'consistent'
+        else:
+            self.report['dns_consistency'] = 'inconsistent'
+        tcp_connect = self.compare_tcp_experiments()
+
+        got_expected_web_page = (
+            (self.report['body_length_match'] is True or
+             self.report['headers_match'] is True)
+            and self.report['status_code_match'] is True
+        )
+
+        if (dns_consistent == True and tcp_connect == False and
+                experiment_http_failure is not None):
+            blocking = 'tcp_ip'
+
+        # XXX we may want to have different codes for these two types of
+        # blocking
```

Comment:
Sounds reasonable, then maybe one can be `http-blockpage` and the other can be `http-failure`?